sdk/python/kfp/dsl/_pipeline.py

```diff
@@ -82,31 +90,16 @@ class Pipeline():
   # _default_pipeline is set when it (usually a compiler) runs "with Pipeline()"
   _default_pipeline = None
 
-  # All pipeline functions with @pipeline decorator that are imported.
-  # Each key is a pipeline function. Each value is a (name, description).
-  _pipeline_functions = []
-
   @staticmethod
   def get_default_pipeline():
     """Get default pipeline. """
     return Pipeline._default_pipeline
 
-  @staticmethod
-  def get_pipeline_functions():
-    """Get all imported pipeline functions (decorated with @pipeline)."""
-    return Pipeline._pipeline_functions
-
-  @staticmethod
-  def _add_pipeline_to_global_list(func):
-    """Add a pipeline function (decorated with @pipeline)."""
-    Pipeline._pipeline_functions.append(func)
-
   @staticmethod
   def add_pipeline(name, description, func):
```

Comment:
maybe we could remove the add_pipeline function here if we instruct the users to only use the decorators.

Reply:
Well... we did instruct the users.
In 3+ places.
We even raised exceptions in the compiler.
Unfortunately, Ajay has decided to use this function in TFX code. See https://github.com/tensorflow/tfx/pull/39 for details.
I've prepared the fix, but now we cannot just delete this function - we have to deprecate it over time =(

Reply:
Make sense. If the TFX sample is using this function, let's keep it for now.

Reply:
Decorators are great for users when constructing a pipeline. However, decorators are syntactic sugar, and ideally, we should also provide a public-facing API to achieve the same result.  Consider a power-user, who wants to build 100 different pipelines, each with slightly different parameters/components within it. By removing this method, we make it not possible for the user to do this. I'm all for cleaning the code up and not exposing internals, but I think we should offer a public API for users to programmatically build pipelines. Right now, we seem to be removing public (by definition) API methods, which is a little strange to me.

What do you think?
/cc @paveldournov 
/cc @vicaire 
/cc @IronPan

Reply:
>Consider a power-user, who wants to build 100 different pipelines, each with slightly different parameters/components within it.
>By removing this method, we make it not possible for the user to do this.

That's not the case. Why do you think so?
You can have hundreds of pipelines. You do not need that undocumented function for that.

You can apply the decorator to 100s of pipelines.
Also, decorator is just a function. You can call a function on 100s of pipelines.
I see no problem.
And I see no reason why someone would want to circumvent it.
There might be some misunderstanding on how the decorators and DSL compiler work. I'll be happy to clear those misunderstandings.

Reply:
Also, as you see, I'm not removing this method now. That's unfortunate, but that TFX code was released and we now have to support it for a while. I wish our team could review it before the release.